validator/rpc/handlers_keymanager.go

```diff
@@ -0,0 +1,92 @@
+package rpc
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/rpc/eth/shared"
+	fieldparams "github.com/prysmaticlabs/prysm/v4/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	http2 "github.com/prysmaticlabs/prysm/v4/network/http"
+	"github.com/prysmaticlabs/prysm/v4/validator/client"
+	"go.opencensus.io/trace"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+// SetVoluntaryExit creates a signed voluntary exit message and returns a VoluntaryExit object.
+func (s *Server) SetVoluntaryExit(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "validator.keymanagerAPI.SetVoluntaryExit")
+	defer span.End()
+
+	rawPubkey := mux.Vars(r)["pubkey"]
+	if rawPubkey == "" {
+		http2.HandleError(w, "pubkey is required in URL params", http.StatusBadRequest)
+		return
+	}
+
+	pubkey, valid := shared.ValidateHex(w, "pubkey", rawPubkey, fieldparams.BLSPubkeyLength)
+	if !valid {
+		return
+	}
+
+	var epoch primitives.Epoch
+	ok, _, e := shared.UintFromQuery(w, r, "epoch")
+	if !ok {
+		http2.HandleError(w, "Invalid epoch", http.StatusBadRequest)
+		return
+	}
+	epoch = primitives.Epoch(e)
+
+	if s.validatorService == nil {
+		http2.HandleError(w, "Validator service not ready", http.StatusPreconditionRequired)
+		return
+	}
+
+	if s.wallet == nil {
+		http2.HandleError(w, "No wallet found", http.StatusPreconditionRequired)
+		return
+	}
+	km, err := s.validatorService.Keymanager()
+	if err != nil {
+		http2.HandleError(w, err.Error(), http.StatusPreconditionRequired)
+		return
+	}
```

Comment:
I would move these checks to the beginning because it doesn't make sense to start handling the request when these preconditions are not satisfied.